{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explanation of tree modification code\n",
    "\n",
    "### Sections of this notebook\n",
    "**How and Why** - explanations of what we used this code for and the cases of modifications made.  \n",
    "**Overall code flow** - the general steps of the code.  \n",
    "**File explanations** - output filenames and what they represent.  \n",
    "**Visualization explanation** - this code produces a modified tree topology with modified names representing strains for computations and visualizations.  The names also undergo important modifications as explained in this section.   \n",
    "**Code and tests** - The actual workflow and tests for critical functions.  \n",
    "\n",
    "\n",
    "### How and Why\n",
    "For strains in the database we generated short amplicon reads of the V4 regions of the 16s rRNA (primers: 515f, 806r). Many of the strains in our collection are identical over this range, leading to a single 16s feature representing many metabolically/physiologically distinct entities. \n",
    "\n",
    "For many of the analyses we perform, we need a phylogenetic tree that has an individual tip for each distinct strain; a 1-1 relationship between strains and tips in the tree. To build this, we first construct the set of all subgraphs (as described below). We then take these subgraphs, and modify the phylogenetic tree reconstructed from the 16s features until it has an exact 1-1 correspondence between tips and strains.\n",
    "\n",
    "For the purposes of this code, imagine a bipartite graph with a strain as one node type and a 16s features as another node type. Edges in this graph connect strain nodes to the 16s features in their genomes (nodes). Starting at a strain  (s) we want to find the subgraph that includes s. In most cases, a strain has a single, unique 16s feature. So the graph just includes s and its 16s feature (e.g. (s, feature_a)). There are three other cases that we must deal with as detailed in the trees below.\n",
    "\n",
    "\n",
    "Case 1: a strain has multiple 16s features in the V4 region. In other words, the subgraph contains one strain node and several feature nodes. In this example, `strain3` has two 16s features. To create a phylogeny where there is a 1-1 correspondence between tips and strains, we need to remove tree tips C and D, and replace them with a new tip. The distance between this tip and it's parent is set to the average of distance between C and the parent and D and the parent. \n",
    "\n",
    "The table of strain to feature correspondence.\n",
    "\n",
    "strain\t16s_feature  \n",
    "strain1\tA  \n",
    "strain2\tB  \n",
    "strain3\tC,D  \n",
    "\n",
    "Tree string: ((A:0.1, B:0.1)in1:0.2,(C:0.1, D:0.3)in2:0.2)in3:0.3;  \n",
    "```\n",
    "                    /-A\n",
    "          /in1-----|\n",
    "         |          \\-B\n",
    "-in3-----|\n",
    "         |          /-C\n",
    "          \\in2-----|\n",
    "                    \\-D\n",
    "```\n",
    "The tree string and visualization after we've combined tips C and D into new tip E.\n",
    "\n",
    "strain\t16s_feature    \n",
    "strain1\tA  \n",
    "strain2\tB  \n",
    "strain3\tE  \n",
    "\n",
    "After filtration.\n",
    "Tree string: ((A:0.1, B:0.1)in1:0.2,E:0.4)in3:0.3;  \n",
    "```\n",
    "                    /-A\n",
    "          /in1-----|\n",
    "-in3-----|          \\-B\n",
    "         |\n",
    "          \\-E\n",
    "```\n",
    "\n",
    "There is a more complex version of this case where the last common ancestor of the tips to be combined also has descendants that shouldn't be combined. In the follow example, we want to combine cf137, cf138, cf139, and cf140 into a tip called 'NEW'. However, the LCA of these tips (in121) contains anotehr descendant tip that we can't remove.\n",
    "\n",
    "strain\t16s_feature   \n",
    "strain1\tcf137,cf138,cf139,cf140  \n",
    "strain2\tcfXXX  \n",
    "strain3\tcf134  \n",
    "strain4\tcf135  \n",
    "strain5\tcf136  \n",
    "\n",
    "Tree string: ((cf137:0.00619,((cf138:0.00394,cf139:0.0,cfXXX:0.4)in123:0.0,cf140:0.00388)in122:0.00567)in121:0.04868,(cf135:0.00218,(cf136:0.00325,cf134:0.0007)in125:0.00177)in124:0.03926)in120:0.00239;   \n",
    "```\n",
    "                    /-cf137\n",
    "                   |\n",
    "                   |                    /-cf138\n",
    "          /in121---|                   |\n",
    "         |         |          /in123---|--cf139\n",
    "         |         |         |         |\n",
    "         |          \\in122---|          \\-cfXXX\n",
    "-in120---|                   |\n",
    "         |                    \\-cf140\n",
    "         |\n",
    "         |          /-cf135\n",
    "          \\in124---|\n",
    "                   |          /-cf136\n",
    "                    \\in125---|\n",
    "                              \\-cf134\n",
    "```\n",
    "After filtration. \n",
    "Tree string: ((cfXXX:0.400567, NEW:0.007755)in121:0.04868,(cf135:0.00218,(cf136:0.00325,cf134:0.0007)in125:0.00177)in124:0.03926)in120:0.00239;\n",
    "```\n",
    "                    /-cfXXX\n",
    "          /in121---|\n",
    "         |          \\-NEW\n",
    "-in120---|\n",
    "         |          /-cf135\n",
    "          \\in124---|\n",
    "                   |          /-cf136\n",
    "                    \\in125---|\n",
    "                              \\-cf134\n",
    "``` \n",
    "\n",
    "Case 2: multiple strains share the same 16s feature. In this subgraph, there is one 16s node and multiple strain nodes connected to it. In this example, `strain1` through `strain4` share the same 16s feature. Here, we convert the existing tip to an internal node by adding four 0-length tips. These allow us to have a 1-1 correspondence between strains and tips without introducing any inaccuracies in branchlength. \n",
    "\n",
    "The table of strain to feature correspondence.  \n",
    "strain\t16s_feature  \n",
    "strain1\tA    \n",
    "strain2\tA  \n",
    "strain3\tA  \n",
    "strain4\tA  \n",
    "strain5 B  \n",
    "strain6\tC\t  \n",
    "strain7\tD  \n",
    "\n",
    "The tree string in Newick format and a visualization (without branch length).  \n",
    "((A:0.1, B:0.1)in1:0.2,(C:0.1, D:0.3)in2:0.2)in3:0.3;\n",
    "```\n",
    "                    /-A\n",
    "          /in1-----|\n",
    "         |          \\-B\n",
    "-in3-----|\n",
    "         |          /-C\n",
    "          \\in2-----|\n",
    "                    \\-D\n",
    "```\n",
    "We n-furcate tip A into four 0-length tips.   \n",
    "(((A1:0, A2:0, A3:0, A4:0)A:0.1, B:0.1)in1:0.2,(C:0.1, D:0.3)in2:0.2)in3:0.3;  \n",
    "The resulting tree looks like the following (again with branchlengths not accurately represented).  \n",
    "```\n",
    "                              /-A1\n",
    "                             |\n",
    "                             |--A2\n",
    "                    /A-------|\n",
    "                   |         |--A3\n",
    "          /in1-----|         |\n",
    "         |         |          \\-A4\n",
    "         |         |\n",
    "-in3-----|          \\-B\n",
    "         |\n",
    "         |          /-C\n",
    "          \\in2-----|\n",
    "                    \\-D\n",
    "\n",
    "```\n",
    "Case 3: All of the remaining cases can be thought of in the same way. The subgraph contains some number of strains and 16s feature nodes in any of a number of relationships. In this example (which is common), we have three related strains. `strain1` has a single feature (A). `strain2` shares feature A with `strain1` and has another feature in its genome (B). `strain3` shares no features with `strain1` but shares a features with `strain2` and thus a path to `strain1`. To resolve situations like these, we remove all tips (16s features) found in this subgraph. We find their last-common-ancestor internal node and find the average length from the deleted tips to that LCA node. We add a new branch from that internal node (of the just calculated length) and then new 0-length tips (as many as there are strains in the subgraph). This introduces some topological and distance inaccuracy, but we felt was the best way to capture the complex relationship between single 16s features and the strains they represent. \n",
    "\n",
    "The table of strain to feature correspondence.\n",
    "strain\t16s_feature  \n",
    "strain1\tA  \n",
    "strain2\tA,B  \n",
    "srrain3\tA,D  \n",
    "strain4\tD  \n",
    "strain5 C  \n",
    "\t\n",
    "\n",
    "The tree string in Newick format and a visualization  \n",
    "(((A:0.1, B:0.1)in1:0.1, C:0.1)in3:0.1, D:0.4)in4:0.1;\n",
    "\n",
    "```\n",
    "                              /-A\n",
    "                    /in1-----|\n",
    "          /in3-----|          \\-B\n",
    "         |         |\n",
    "-in4-----|          \\-C\n",
    "         |\n",
    "          \\-D\n",
    "```\n",
    "\n",
    "Now, we must remove features A, B, and D, and build a new tree. Note that we leave C unchanged because it's not part of the subgraph of the A,B,D features. Also note the length from   \n",
    "((n1:0, n2:0, n3:0, n4:0)inX:0.25, C:0.1)in4:0.1;\n",
    "\n",
    "```\n",
    "                    /-n1\n",
    "                   |\n",
    "                   |--n2\n",
    "          /inX-----|\n",
    "         |         |--n3\n",
    "-in4-----|         |\n",
    "         |          \\-n4\n",
    "         |\n",
    "          \\-C\n",
    "```\n",
    "Note that you can easily verify these examples using scikit-bio.\n",
    "```python\n",
    "from skbio import TreeNode\n",
    "import io\n",
    "f = io.StringIO(\"((n1:0, n2:0, n3:0, n4:0)in1:0.25, C:0.1)in4:0.1;A\")\n",
    "tre = TreeNode.read(f)\n",
    "print(tre.ascii_art())\n",
    "```\n",
    "\n",
    "### Overall code flow\n",
    "1. Build subgraphs of strain/features.\n",
    "2. Modify tree so that toplogy reflects 1-1 relationships using the strain/feature subgraphs.\n",
    "3. Shear the tree so that strains/features that we don't want to show up are removed.\n",
    "4. Create visualizations and files necessary to make iToll compatible.\n",
    "\n",
    "### File explanations\n",
    "**`input_tree.tre`** - The phylogenetic tree that will be modified. Must have branch lengths and internal nodes named.  \n",
    "**`modified_tree.tre`** - This is the tree where all topological changes have been made. Node names have also been updated to reflect strain names. Some nodes are named with modifications of strain names to aid visualization as explained in that section of the code (nodes named with `nfurTip` or `cTip`).\n",
    "**`modified_tree.sheared.tre`** - This tree has all exposed internal nodes removed from `modified_tree.tre`.  \n",
    "**`modified_tree.sheared.unwanted_removed.tre`** - This tree is the same as above, but with additional tips removed as specified in the code (e.g. to produce a Bacteroidetes) only tree.  \n",
    "**`strain_names.modified_tree.sheared.unwanted_removed.tre`** - This tree is the `modified_tree.sheared.tre` but with all tips named as strains. In other words, the `nfur` and `cTip` has been removed from the names.  \n",
    "**`itol_order.txt`** - The order (top to bottom) in which strain's will appear on a dendrogram iTOL tree view from `strain_names.modified_tree.sheared.unwanted_removed.tre`.\n",
    "\n",
    "### Visualization explanations\n",
    "If using iTOL, you will likely want to visualize **`modified_tree.sheared.unwanted_removed.tre`** to see where you should add clade markers (e.g. over 0-length tips). You can then overlay the names as shown in **`strain_names.modified_tree.sheared.unwanted_removed.tre`**.  \n",
    "**Note**: we are introducing some strange tip names like `nfurTip__Desulfovibrio piger ATCC 29098 M2` and `cTip__in45__Bacteroides vulgatus ATCC 8482`. We are introducing `nfurTip` into a strain name every time we have a single 16s feature that maps to multiple strains. We are introducing `cTip` into a every time we combine a set of 16s features because they all represent a subgraph. The reason for doing this is so that we can visualize the changes on an iTOL tree. These names appear in `modified_tree.sheared.unwanted_removed.tre` but not in ``strain_names.modified_tree.sheared.unwanted_removed.tre``.\n",
    "**Note 2**: iTOL orders the nodes of a tree in an unusual way (in the normal dendrogram layout). To have a preorder traversal layout you must select 'Advanced: Leaf Sorting: None'. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Imports\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from skbio import TreeNode\n",
    "import io\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code for generating modified tree\n",
    "\n",
    "The starting point is two dicts mapping strain names to the names of the 16s features representing those strains. The 16s feature names must be the names found in the phylogenetic tree.\n",
    "\n",
    "```python\n",
    "taxonomy_to_features = {'Peptostreptococcus sp. CC14N BEI HM-1051': ['cf095', 'cf096'],\n",
    "                        'Eubacterium rectale ATCC 33656': ['cf143'],\n",
    "                        ...}\n",
    "features_to_taxonomy = {'cf136': ['Eubacterium eligens DSMZ 3376'],\n",
    "                        'cf137': ['Anaerobutyricum hallii DSM 3353'],\n",
    "                        'cf138': ['Anaerobutyricum hallii DSM 3353'],\n",
    "                        ...}\n",
    "```\n",
    "\n",
    "Note that if you wanted to modify the tree for only a subset of the tips you could reduce these two dicts to just the strains/tips of interst."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Ingest culture and strain data to produce required dicts. This code is required because of\n",
    "# the specific formatting of our strain and culture database. You should bypass/modify\n",
    "# depending on the way you store this information.\n",
    "strains_and_cultures_db = '../supplemental_table_6.xlsx'\n",
    "sac = pd.read_excel(strains_and_cultures_db, index_col=0, sheet_name='cultures')\n",
    "taxonomies = pd.read_excel(strains_and_cultures_db, index_col=0, sheet_name='full_taxonomy')\n",
    "\n",
    "# Join taxonomy data with sac to link culture to taxonomic rank information\n",
    "tmp = ['taxonomy', 'clean_by_16s', '16s_feature', 'kingdom', 'phylum', 'class',\n",
    "       'order', 'family', 'genus', 'species', 'strain', 'morphology',\n",
    "       'supplier', 'number']\n",
    "strain_metadata = sac.join(taxonomies, on='taxonomy')[tmp]\n",
    "\n",
    "# Build dict's we will use to find subgraph as described above.\n",
    "taxonomy_to_cf = defaultdict(list)\n",
    "cf_to_taxonomy = defaultdict(list)\n",
    "\n",
    "for t, cf in strain_metadata[['taxonomy', '16s_feature']].values:\n",
    "    if pd.isnull(cf):\n",
    "        pass\n",
    "    else:\n",
    "        taxonomy_to_cf[t].extend(cf.split(','))\n",
    "        for _cf in cf.split(','):\n",
    "            cf_to_taxonomy[_cf].append(t)\n",
    "\n",
    "t2c = {k: sorted(set(v)) for k, v in taxonomy_to_cf.items()}\n",
    "c2t = {k: sorted(set(v)) for k, v in cf_to_taxonomy.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build a new tree (if necessary)\n",
    "We need to align sequences from our cultures and build a tree. This step can be skipped if you've alread built the tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "seqs = pd.read_excel(strains_and_cultures_db, index_col=0, sheet_name='v4_16s')\n",
    "seq_fasta = []\n",
    "for cn, seq in zip(seqs.index, seqs['sequence'].values):\n",
    "    seq_fasta.append('>%s\\n%s' % (cn, seq))\n",
    "out_fp = 'seqs_to_align.fasta'\n",
    "o = open(out_fp, 'w')\n",
    "o.writelines('\\n'.join(seq_fasta))\n",
    "o.close()\n",
    "# Now align with whatever program. Here we'll use CLUSTALW via the MUSCLE pipeline at EBI,"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now you will need to build the tree with these sequences. The next blocks of code assume you've created a tree, which I've called `clustalw_tree.tre`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Name internal nodes of the tree\n",
    "This is required to find nodes and last common ancestors. If a tree already has internal node names, this step is unnecessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tree_fp = 'clustalw_tree.tre'\n",
    "tr = TreeNode.read(tree_fp, format='newick')\n",
    "\n",
    "i = 0\n",
    "for node in tr.preorder():\n",
    "    if node.name == None:\n",
    "        node.name = 'in%s' % i\n",
    "        i += 1\n",
    "out_fp = 'base_tree.named_inodes.tre'\n",
    "tr.write(out_fp, format='newick')\n",
    "tr = TreeNode.read('base_tree.named_inodes.tre', format='newick')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build subgraphs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def strains_from_features(features_16s, feature_to_strain):\n",
    "    tmp = []\n",
    "    for f in features_16s:\n",
    "        tmp.extend(feature_to_strain[f])\n",
    "    return sorted(set(tmp))\n",
    "\n",
    "def features_from_strains(strains, strain_to_feature):\n",
    "    tmp = []\n",
    "    for s in strains:\n",
    "        tmp.extend(strain_to_feature[s])\n",
    "    return sorted(set(tmp))\n",
    "\n",
    "def find_subgraph(start_features, feature_to_strain, strain_to_feature):\n",
    "    _16s_features = start_features\n",
    "    _strains = strains_from_features(start_features, feature_to_strain)\n",
    "    subgraph = False\n",
    "    while not subgraph:\n",
    "        tmp1 = features_from_strains(_strains, strain_to_feature)\n",
    "        tmp2 = strains_from_features(tmp1, feature_to_strain)\n",
    "        if (tmp1 == _16s_features) and (tmp2 == _strains):\n",
    "            subgraph = True\n",
    "        else:\n",
    "            subgraph = False\n",
    "        _16s_features = tmp1\n",
    "        _strains = tmp2\n",
    "\n",
    "    return (_16s_features, _strains)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test code\n",
    "Ensure that we are finding subgraphs like expected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "feature_to_strain_1 = {'cf01': ['bt1', 'bt2'],\n",
    "                       'cf02': ['cs1'],\n",
    "                       'cf03': ['am1', 'am2'],\n",
    "                       'cf04': ['am2'],\n",
    "                       'cf05': ['fp1', 'fp2'],\n",
    "                       'cf06': ['fp1', 'fp2']}\n",
    "\n",
    "strain_to_feature_1 = {'bt1': ['cf01'],\n",
    "                       'bt2': ['cf01'],\n",
    "                       'cs1': ['cf02'],\n",
    "                       'am1': ['cf03'],\n",
    "                       'am2': ['cf03', 'cf04'],\n",
    "                       'fp1': ['cf05', 'cf06'],\n",
    "                       'fp2': ['cf05', 'cf06']}\n",
    "\n",
    "obs1, obs2 = find_subgraph(['cf01'], feature_to_strain_1, strain_to_feature_1)\n",
    "assert obs1 == ['cf01']\n",
    "assert obs2 == ['bt1', 'bt2']\n",
    "\n",
    "obs1, obs2 = find_subgraph(['cf02'], feature_to_strain_1, strain_to_feature_1)\n",
    "assert obs1 == ['cf02']\n",
    "assert obs2 == ['cs1']\n",
    "\n",
    "obs1, obs2 = find_subgraph(['cf03'], feature_to_strain_1, strain_to_feature_1)\n",
    "assert obs1 == ['cf03', 'cf04']\n",
    "assert obs2 == ['am1', 'am2']\n",
    "\n",
    "obs1, obs2 = find_subgraph(['cf04'], feature_to_strain_1, strain_to_feature_1)\n",
    "assert obs1 == ['cf03', 'cf04']\n",
    "assert obs2 == ['am1', 'am2']\n",
    "\n",
    "obs1, obs2 = find_subgraph(['cf05'], feature_to_strain_1, strain_to_feature_1)\n",
    "assert obs1 == ['cf05', 'cf06']\n",
    "assert obs2 == ['fp1', 'fp2']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create subgraphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Create graphs from bipartite network defined by `t2c` and `c2t`.\n",
    "remaining_16s_features = sorted(c2t.keys())\n",
    "subgraphs = []\n",
    "while len(remaining_16s_features) > 0:\n",
    "    _16s_features, _strains = find_subgraph([remaining_16s_features[0]], c2t, t2c)\n",
    "    for i in _16s_features:\n",
    "        remaining_16s_features.pop(remaining_16s_features.index(i))\n",
    "    subgraphs.append((_16s_features, _strains))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Modify the tree using identified subgraphs\n",
    "#### Define functions to modify a `skbio.TreeNode`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def prune(tr, tips_to_remove):\n",
    "    ntr = tr.copy()\n",
    "    # get lca before removal\n",
    "    lca_name = ntr.lca(tips_to_remove).name\n",
    "    # find average distance from lca to tips to be removed\n",
    "    bl = 0.\n",
    "    for tip in tips_to_remove:\n",
    "        bl += ntr.find(tip).accumulate_to_ancestor(ntr.find(lca_name))\n",
    "    avg_bl = bl / len(tips_to_remove)\n",
    "    # remove tips\n",
    "    ntr.remove_deleted(lambda node: node.name in set(tips_to_remove))\n",
    "    return ntr, lca_name, avg_bl\n",
    "\n",
    "def prune_and_combine_into_1(tr, tips_to_remove, new_tip_name):\n",
    "    ntr = tr.copy()\n",
    "    ntr_1, lca_name, avg_bl = prune(tr, tips_to_remove)\n",
    "    # In some topologies, after tip removal, the lca node will not be a tip, but will \n",
    "    # have descendant internal nodes and tips. We have two cases that differ in how we\n",
    "    # set branch length.\n",
    "    if set([tip.name for tip in tr.find(lca_name).tips()]) == set(tips_to_remove):\n",
    "        lca_bl = ntr_1.find(lca_name).length\n",
    "        ntr_1.find(lca_name).parent.children.append(TreeNode(name=new_tip_name, length=avg_bl + lca_bl))\n",
    "        # Delete the previous LCA to remove every internal node that might still be attached.\n",
    "        ntr_1.remove_deleted(lambda node: node.name == lca_name)\n",
    "    else:\n",
    "        # Here we have other tips that descend from the LCA that aren't getting cleaned up. Add new node\n",
    "        # with appropriate branch length.\n",
    "        ntr_1.find(lca_name).children.append(TreeNode(name=new_tip_name, length=avg_bl))\n",
    "        # Prune away the possibly meaningless internal nodes.\n",
    "        ntr_1.prune()\n",
    "    return ntr_1\n",
    "\n",
    "def nfurcate_tip(tr, tip, tips_to_add):\n",
    "    ntr = tr.copy()\n",
    "    for tip_name in tips_to_add:\n",
    "        ntr.find(tip).children.append(TreeNode(name=tip_name, length=0.0))\n",
    "    return ntr\n",
    "\n",
    "def add_0_length_tips(tr, lca_name, tips_to_add):\n",
    "    ntr = tr.copy()\n",
    "    for tip_name in tips_to_add:\n",
    "        ntr.find(lca_name).children.append(TreeNode(name=tip_name, length=0.0))\n",
    "    return ntr\n",
    "\n",
    "def make_clade(tr, tips_to_remove, tips_to_add, new_inode_name):\n",
    "    ntr = tr.copy()\n",
    "    ntr_1, lca_name, avg_bl = prune(ntr, tips_to_remove)\n",
    "    # Add the new internal node that will be parent to the new tips.\n",
    "    ntr_1.find(lca_name).children.append(TreeNode(name=new_inode_name, length=avg_bl))\n",
    "    ntr_2 = add_0_length_tips(ntr_1, new_inode_name, tips_to_add)\n",
    "    return ntr_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test code\n",
    "Ensure our functions modify topology and length as we expect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tr_str1 = \"((A:0.1, B:0.34)in1:0.2,(C:0.1, D:0.3)in2:0.2)in3:0.3;\"\n",
    "tr_str2 = \"(((A1:0, A2:0, A3:0, A4:0)A:0.5, B:0.1)in1:0.2,(C:0.1, D:0.3)in2:0.2)in3:0.3;\"\n",
    "\n",
    "tr_str3 = \"((C:0.1, D:0.1)B:0.1, (F:0.1, G:0.1)E:0.1)A:0.1;\"\n",
    "exp_tr_str3 = \"(B:0.1, (N1:0.0, N2:0.0, N3:0.0)n_inode:0.2, (F:0.1)E:0.1)A:0.1;\"\n",
    "\n",
    "tr_str4 = \"(((C:0.1, D:0.1)B:0.1, (F:0.1, G:0.1)E:0.1)A:0.1, (X1:0.1, X2:0.1)X3:0.1)root:0.1;\"\n",
    "exp_tr_str4 = \"((B:0.1, (N1:0.0, N2:0.0, N3:0.0)n_inode:0.2, (F:0.1)E:0.1)A:0.1, (X1:0.1, X2:0.1)X3:0.1)root:0.1;\"\n",
    "\n",
    "tr_str5 = \"((cf137:0.00619,((cf138:0.00394,cf139:0.0)in123:0.0,cf140:0.00388)in122:0.00567)in121:0.04868,(cf135:0.00218,(cf136:0.00325,cf134:0.0007)in125:0.00177)in124:0.03926)in120:0.00239;\"\n",
    "exp_tr_str5 = \"(NEW:0.056435,(cf135:0.00218,(cf136:0.00325,cf134:0.0007)in125:0.00177)in124:0.03926)in120:0.00239;\"\n",
    "\n",
    "tr_str6 = \"((cf137:0.00619,((cf138:0.00394,cf139:0.0,cfXXX:0.4)in123:0.0,cf140:0.00388)in122:0.00567)in121:0.04868,(cf135:0.00218,(cf136:0.00325,cf134:0.0007)in125:0.00177)in124:0.03926)in120:0.00239;\"\n",
    "exp_tr_str6 = \"((cfXXX:0.400567, NEW:0.007755)in121:0.04868,(cf135:0.00218,(cf136:0.00325,cf134:0.0007)in125:0.00177)in124:0.03926)in120:0.00239;\"\n",
    "\n",
    "# prune\n",
    "_test_tr = TreeNode.read(io.StringIO(tr_str1))\n",
    "obs_tr, obs_lca, obs_bl = prune(_test_tr, ['A', 'B'])\n",
    "assert obs_lca == 'in1'\n",
    "np.testing.assert_allclose(obs_bl, 0.22)\n",
    "\n",
    "_test_tr = TreeNode.read(io.StringIO(tr_str2))\n",
    "obs_tr, obs_lca, obs_bl = prune(_test_tr, ['A1', 'A2'])\n",
    "assert obs_lca == 'A'\n",
    "np.testing.assert_allclose(obs_bl, 0.0)\n",
    "\n",
    "obs_tr, obs_lca, obs_bl = prune(_test_tr, ['A1', 'A2', 'C'])\n",
    "assert obs_lca == 'in3'\n",
    "np.testing.assert_allclose(obs_bl, 1.7/3)\n",
    "\n",
    "# make_clade\n",
    "_test_tr = TreeNode.read(io.StringIO(tr_str3))\n",
    "obs_tr = make_clade(_test_tr, ['C', 'D', 'G'], ['N1', 'N2', 'N3'], 'n_inode')\n",
    "exp_tr = TreeNode.read(io.StringIO(exp_tr_str3))\n",
    "assert obs_tr.compare_rfd(exp_tr) == 0.0\n",
    "assert obs_tr.compare_tip_distances(exp_tr) == 0.0\n",
    "\n",
    "_test_tr = TreeNode.read(io.StringIO(tr_str4))\n",
    "obs_tr = make_clade(_test_tr, ['C', 'D', 'G'], ['N1', 'N2', 'N3'], 'n_inode')\n",
    "exp_tr = TreeNode.read(io.StringIO(exp_tr_str4))\n",
    "assert obs_tr.compare_rfd(exp_tr) == 0.0\n",
    "assert obs_tr.compare_tip_distances(exp_tr) == 0.0\n",
    "\n",
    "# prune_and_combine_into_1\n",
    "_test_tr = TreeNode.read(io.StringIO(tr_str5))\n",
    "obs_tr = prune_and_combine_into_1(_test_tr, ['cf137', 'cf138', 'cf139', 'cf140'], 'NEW')\n",
    "exp_tr = TreeNode.read(io.StringIO(exp_tr_str5))\n",
    "# NEW_bl = 0.04868 + 0.007755\n",
    "#avg_bl = 0.007755\n",
    "#(in122 + in123 + cf138) + (in122 + in123 + cf139) + (in122 + cf140) + (cf137)\n",
    "#(0.00567 + 0 + 0.00394) + (0.00567 + 0 + 0) + (0.00567 + 0.00388) + (0.00619)\n",
    "assert obs_tr.compare_rfd(exp_tr) == 0.0\n",
    "assert obs_tr.compare_tip_distances(exp_tr) == 0.0\n",
    "assert obs_tr.find('NEW').length == 0.056435\n",
    "\n",
    "_test_tr = TreeNode.read(io.StringIO(tr_str6))\n",
    "obs_tr = prune_and_combine_into_1(_test_tr, ['cf137', 'cf138', 'cf139', 'cf140'], 'NEW')\n",
    "exp_tr = TreeNode.read(io.StringIO(exp_tr_str6))\n",
    "assert obs_tr.compare_rfd(exp_tr) == 0.0\n",
    "np.testing.assert_allclose(obs_tr.compare_tip_distances(exp_tr), 0.0, rtol=0, atol=1e-5)\n",
    "np.testing.assert_allclose(obs_tr.find('NEW').length, 0.007755)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Modify tree\n",
    "This code modifies the topology of the tree to resolve the three cases outlined in explanation. It also updates the names of the tips so that they match the supplied strain names. This is required for visualization and because nodes need unique names in an `skbio.TreeNode`.\n",
    "\n",
    "**Note**: we are introducing some strange tip names like `nfurTip__Desulfovibrio piger ATCC 29098 M2` and `cTip__in45__Bacteroides vulgatus ATCC 8482`. See explanation in **Visualization explanation** in first cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Load tree\n",
    "base_tree_fp = 'base_tree.named_inodes.tre'\n",
    "tr = TreeNode.read(base_tree_fp, format='newick')\n",
    "\n",
    "# Keep track of the tips we want to leave in the final tree. This tracks the changes we've\n",
    "# made\n",
    "tips_to_keep = []\n",
    "changes = []\n",
    "# [case, _16s_features, _strains, removed_tipnames, added_tipnames]\n",
    "\n",
    "# Keep track of the new internal node names that are being used.\n",
    "new_inode_counter = 0\n",
    "\n",
    "for _16s_features, _strains in subgraphs:\n",
    "    n_16s_features = len(_16s_features)\n",
    "    n_strains = len(_strains)\n",
    "    \n",
    "    if (n_16s_features == 1) and (n_strains == 1):\n",
    "        # No editing needs to be done.\n",
    "        changes.append((0, _16s_features, _strains, _16s_features, _strains))\n",
    "        \n",
    "        #rename.\n",
    "        tr.find(_16s_features[0]).name = _strains[0]\n",
    "        tips_to_keep.append(_strains[0])\n",
    "\n",
    "    elif (n_16s_features > 1) and (n_strains == 1):\n",
    "        # Case 1\n",
    "        tr = prune_and_combine_into_1(tr, _16s_features, _strains[0])\n",
    "        tips_to_keep.append(_strains[0])\n",
    "\n",
    "        changes.append((1, _16s_features, _strains, _16s_features, _strains[0]))\n",
    "        \n",
    "    elif (n_16s_features == 1) and (n_strains > 1):\n",
    "        # Case 2\n",
    "        new_names = ['nfurTip__%s' % i for i in _strains]\n",
    "        tr = nfurcate_tip(tr, _16s_features[0], new_names)\n",
    "        # Add new tips to keep.\n",
    "        tips_to_keep.extend(new_names)\n",
    "        changes.append((2, _16s_features, _strains, _16s_features, new_names))\n",
    "\n",
    "    else:\n",
    "        # Case 3\n",
    "        new_tip_names = ['cTip__%s__%s' % (tr.lca(_16s_features).name, i) for i in \n",
    "                         _strains]\n",
    "        new_inode_name = 'new_inode_' + str(new_inode_counter).zfill(3)\n",
    "        new_inode_counter += 1\n",
    "        tr = make_clade(tr, _16s_features, new_tip_names, new_inode_name)\n",
    "\n",
    "        tips_to_keep.extend(new_tip_names)\n",
    "\n",
    "\n",
    "        changes.append((2, _16s_features, _strains, _16s_features, new_tip_names))\n",
    "\n",
    "# Write the resulting file.\n",
    "out_fp = 'modified_tree.tre'\n",
    "tr.write(out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make final modifications to tree for visualization\n",
    "#### Shearing\n",
    "Because of the topology changes, some internal nodes have been exposed as tips. We need to remove these and shear them back until we reach a node with an LCA tip with a tip we want."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Sanity check that all tips we want to keep are in the tree\n",
    "tips = [i.name for i in tr.tips()]\n",
    "assert len(set(tips_to_keep).difference(tips)) == 0\n",
    "# Shear.\n",
    "out_fp = 'modified_tree.sheared.tre'\n",
    "sheared_tr = tr.shear(tips_to_keep)\n",
    "sheared_tr.write(out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Remove and rename tips, and reshear\n",
    "In some cases, we may only want a subset of the tips (e.g. make a tree for visualization with only the Bacteroidetes). Here we will translate the tip names to those we want and produce twin files with things named only as strains, and those named with e.g. 'cTip' to add our visualization in iTOL.\n",
    "\n",
    "**Note**: for some reason, certain tips show up with quotation marks upon insertion. I've removed these by hand with simple find and replace in the final files to be visualized."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# You could substitue any list of tips here. `tips_to_keep` is the set of strain names we want to retain.\n",
    "data_fp = '../supplemental_table_7.xlsx'\n",
    "j_agg_md = pd.read_excel(io=data_fp, sheet_name='agggregated_md', index_col=0)\n",
    "mega_media_samples = ((j_agg_md['sample_type'] == 'supernatant') &\n",
    "                      (j_agg_md['media'] == 'mm'))\n",
    "ss_md = j_agg_md.loc[mega_media_samples, :]\n",
    "tips_to_retain = list(set(ss_md['taxonomy'].values))\n",
    "\n",
    "# _tmp_tr1 will be the modified tree with unwanted tips removed but with\n",
    "# the names retained (e.g. there will 'nfurTip' etc in there).\n",
    "_tmp_tr1 = sheared_tr.copy() \n",
    "_tmp_tr1_tips_to_keep = []\n",
    "\n",
    "# _tmp_tr2 will have the unwanted tips removed, and everything renamed.\n",
    "_tmp_tr2 = sheared_tr.copy() \n",
    "\n",
    "name_map = {}\n",
    "\n",
    "for tip in sheared_tr.tips():\n",
    "    existing_name = tip.name\n",
    "    if 'nfurTip__' in existing_name:\n",
    "        new_name = existing_name.replace('nfurTip__', '')\n",
    "    elif 'cTip' in existing_name:\n",
    "        new_name = re.sub('cTip__in[0-9]+__', '', existing_name)\n",
    "    else:\n",
    "        new_name = existing_name\n",
    "    # Sanity check\n",
    "    if existing_name in name_map:\n",
    "        raise ValueError('reused name')\n",
    "    \n",
    "    name_map[existing_name] = new_name\n",
    "    \n",
    "    if new_name not in tips_to_retain:\n",
    "        pass\n",
    "    else:\n",
    "        _tmp_tr1_tips_to_keep.append(existing_name)\n",
    "        _tmp_tr2.find(existing_name).name = new_name\n",
    "\n",
    "# Invalidate caches so name changes take effect\n",
    "_tmp_tr2.invalidate_caches()\n",
    "        \n",
    "unwanted_removed = _tmp_tr1.shear(_tmp_tr1_tips_to_keep)\n",
    "out_fp = 'modified_tree.sheared.unwanted_removed.tre'\n",
    "unwanted_removed.write(out_fp)\n",
    "\n",
    "strain_names_unwanted_removed = _tmp_tr2.shear(tips_to_retain)\n",
    "out_fp = 'strain_names.modified_tree.sheared.unwanted_removed.tre'\n",
    "strain_names_unwanted_removed.write(out_fp)\n",
    "\n",
    "# Write out order of the names\n",
    "tip_order = []\n",
    "for node in strain_names_unwanted_removed.preorder():\n",
    "    if node.is_tip():\n",
    "        tip_order.append(node.name)\n",
    "\n",
    "out_fp = 'itol_order.txt'\n",
    "o = open(out_fp, 'w')\n",
    "o.writelines('\\n'.join(tip_order))\n",
    "o.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
